shell/platform/android/io/flutter/embedding/engine/android/FlutterSurfaceView.java

```diff
@@ -0,0 +1,159 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.android;
+
+import android.content.Context;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.util.AttributeSet;
+import android.util.Log;
+import android.view.SurfaceHolder;
+import android.view.SurfaceView;
+
+import java.nio.ByteBuffer;
+
+import io.flutter.embedding.engine.renderer.FlutterRenderer;
+
+/**
+ * Paints a Flutter UI on a {@link android.view.Surface}.
+ *
+ * Consider using a {@link FlutterView}, which utilizes a {@code FlutterSurfaceView}
+ * internally.
+ */
+public class FlutterSurfaceView extends SurfaceView {
+  private static final String TAG = "FlutterSurfaceView";
+
+  private boolean isSurfaceAvailableForRendering = false;
+  private boolean isAttachedToFlutterRenderer = false;
+  @Nullable
+  private FlutterRenderer flutterRenderer;
+
+  // Connects the {@code Surface} beneath this {@code SurfaceView} with Flutter's native code.
+  // Callbacks are received by this Object and then those messages are forwarded to our
+  // FlutterRenderer, and then on to the JNI bridge over to native Flutter code.
+  private final SurfaceHolder.Callback surfaceCallback = new SurfaceHolder.Callback() {
+    @Override
+    public void surfaceCreated(SurfaceHolder holder) {
+      Log.d(TAG, "SurfaceHolder.Callback.surfaceCreated()");
+      isSurfaceAvailableForRendering = true;
+
+      if (isAttachedToFlutterRenderer) {
+        Log.d(TAG, "Already attached to renderer. Notifying of surface creation.");
+        connectSurfaceToRenderer();
+      }
+    }
+
+    @Override
+    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
+      if (isAttachedToFlutterRenderer) {
+        changeSurfaceSize(width, height);
+      }
+    }
+
+    @Override
+    public void surfaceDestroyed(SurfaceHolder holder) {
+      Log.d(TAG, "SurfaceHolder.Callback.surfaceDestroyed()");
+      isSurfaceAvailableForRendering = false;
+
+      if (isAttachedToFlutterRenderer) {
+        disconnectSurfaceFromRenderer();
+      }
+    }
+  };
+
+  /**
+   * Constructs a {@code FlutterSurfaceView} programmatically, without any XML attributes.
+   */
+  public FlutterSurfaceView(Context context) {
+    this(context, null);
+  }
+
+  /**
+   * Constructs a {@code FlutterSurfaceView} in an XML-inflation-compliant manner.
+   */
+  public FlutterSurfaceView(Context context, AttributeSet attrs) {
+    super(context, attrs);
+    init();
+  }
+
+  private void init() {
```

Comment:
optional, subjective: I'd prefer a clearer method name to an implementation comment explaining what this method does. In my opinion implementation comments tend to fall out of sync with the actual state of the code so I prefer avoiding them when possible. `registerSurfaceCallbacks`?

Applies to `TextureView#init` as well.

Reply:
The use of an `init()` method is a standard practice I've used on Android teams for years whereby all constructors eventually lead to an invocation of init().  This keeps constructors focused on interpreting all input while leaving all eventual initialization to to an `init()` method.  I think it has value as a repeated pattern, but also as a pattern I don't know that it's worth declaring this purpose for every init() method...